### PR TITLE
Add PageRank algorithm

### PR DESCRIPTION
**Pull request by BSZKaneki**
## Description

This pull request implements the **PageRank** algorithm under the `graph` module. 

### Algorithm Overview
PageRank is an algorithm used to rank elements in a network based on their structure of incoming links. It computes a probability distribution representing the likelihood that a person randomly clicking on links will arrive at any particular page.

For more details, see the Wikipedia article: https://en.wikipedia.org/wiki/PageRank

### Implementation Details
* **Genericity**: Uses a generic type `Node: Hash + Eq + Clone`, meaning it can rank nodes of any type (e.g., `String`, integer types, or custom hashable structures).
* **Dangling Nodes**: Accurately handles "dangling nodes" (nodes with no outbound edges) by redistributing their rank mass evenly across all nodes in the graph during each iteration.
* **Deduplication**: Handles duplicate/parallel edges between the same source and destination to prevent artificial weight skewing.
* **Convergence**: Performs power iteration with an early exit check using an L1 norm difference calculation against a user-defined threshold.
* **Testing**: Includes 18 tests covering edge cases (such as self-loops, disconnected components, and linear chains) and checking against known analytical results. The entire test suite completes in approximately 0.01s.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran below commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `CONTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

**Comment by BSZKaneki** [comment]
@siriak Hello could you please take a look at this pull request when you have a moment?

**Comment by siriak** [comment]
Please check the comments

File: src/graph/page_rank.rs [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+use std::hash::Hash;
+
+/// Calculates the PageRank for each node in a graph.
+///
+/// The graph is represented as an adjacency list: `HashMap<Node, Vec<Node>>`,
+/// where each key is a source node pointing to a vector of destination nodes.
+///
+/// # Parameters
+/// * `graph` - The adjacency list of the graph.
+/// * `damping_factor` - The probability that a surfer continues clicking links should be in bettwen 0 and 1 (typically 0.85).
```
Comment (siriak) [inline src/graph/page_rank.rs]:
```suggestion
/// * `damping_factor` - The probability that a surfer continues clicking links should be between 0 and 1 (typically 0.85).
```

**Review by siriak — APPROVED** [review]
Looks good, thanks!
